### PR TITLE
feat: add docs for NonNullApi and nullability annotation collision

### PR DESCRIPTION
**Pull request by Lodin**
Resolves vaadin/hilla#641

### CODE REVIEW

File: articles/lit/reference/type-nullability.adoc [hunk 1]
```diff
@@ -134,4 +134,83 @@ export default interface MyBean {
...
+
+There could be a case where several annotations are applied to the same type.
+To resolve this collision, each annotation has its priority; among all applied annotations the one with the highest priority is selected.
+
+* `@NonNullApi` annotation has priority of `10` which makes it the weakest.
+This annotation applies only to types without any nullability annotation.
+* `@Nullable` annotation has priority of `20`.
+It makes any type that affected by `@NonNullApi` annotation nullable.
+* `@Nonnull` annotation has priority of `30` which makes it the strongest.
+It makes any type non-nullable again.
```
Comment (cromoteca) [inline articles/lit/reference/type-nullability.adoc:184]: I'd not disclose our priority numbers, they don't matter in the documentation, they are just an implementation detail. Saying the order is enough.

**Review by cromoteca — CHANGES_REQUESTED** [review]

File: articles/lit/reference/type-nullability.adoc [hunk 2]
```diff
@@ -134,4 +134,83 @@ export default interface MyBean {
...
+
+== Nullability collision
+
+There could be a case where several annotations are applied to the same type.
+To resolve this collision, each annotation has its priority; among all applied annotations the one with the highest priority is selected.
+
+* `@NonNullApi` annotation has the lowest priority.
+This annotation applies only to types without any nullability annotation.
+* `@Nullable` annotation has intermediate priority.
+It makes any type that affected by `@NonNullApi` annotation nullable.
```
Comment (cromoteca) [inline articles/lit/reference/type-nullability.adoc:182]: This sentence is not clear

File: articles/lit/reference/type-nullability.adoc [hunk 3]
```diff
@@ -134,4 +134,83 @@ export default interface MyBean {
...
+* `@Nonnull` annotation has the highest priority.
+It makes any type non-nullable again.
```
Comment (cromoteca) [inline articles/lit/reference/type-nullability.adoc:184]: I can't figure out the "again" case

File: articles/lit/reference/type-nullability.adoc [hunk 4]
```diff
@@ -134,4 +134,83 @@ export default interface MyBean {
...
+.`MyBean.java`
+[source,java]
+----
+public class MyBean {
+    public List<String> list;
+    @Nullable
+    public Map<String, @Nullable Integer> nullableMap;
+    @Nonnull
+    @Nullable
+    public Short forcedNonNullNumber;
```
Comment (cromoteca) [inline articles/lit/reference/type-nullability.adoc:202]: While this is useful to explain priorities, I'd not show it as it makes no sense to put both annotations on the same field. Or at least I'd add a comment to clarify that it is just to show the priority.

**Review by cromoteca — CHANGES_REQUESTED** [review]

File: articles/lit/reference/type-nullability.adoc (hunk 3 above)
Comment (Lodin) [reply articles/lit/reference/type-nullability.adoc:184]: Hmm... Would it be more clear this way?
```
It makes any nullable type (either default or made nullable by `@Nullable` annotation) non-nullable. 
```
or it's better to make it as simple as possible? 
```
It makes any nullable type non-nullable.
```

Comment (cromoteca) [reply articles/lit/reference/type-nullability.adoc:184]: I prefer the latter

File: articles/lit/reference/type-nullability.adoc (hunk 2 above)
Comment (cromoteca) [reply articles/lit/reference/type-nullability.adoc:182]: Reading it again, I'm not sure that the whole collision section is needed: you already clarified at line 147 how to restore nullability under `NonNullApi`, and conflicts between `Nullable` and `Nonnull` are unlikely to happen.

Comment (Lodin) [reply articles/lit/reference/type-nullability.adoc:182]: Hmm. You're right. I guess I just followed the implementation details.

**Review by cromoteca — APPROVED** [review]

**Review by platosha — APPROVED** [review]

**Pull request merged**